test/pkcs12_format_test.c

```diff
@@ -0,0 +1,918 @@
+    "YFVO1chu9A5Xvi75Ob5ZP6Iwu1fRoxMuVXx8atjeAr6e7RDQxXMd6j6xVYEC70jI\n"
+    "HFx6krBY0xlbXaK2VmkCQB4Aap+67kZaxbWfkTPdyZZ1t4fPGBy3uT8EELh1qbig\n"
+    "MTUDMInIN2ggMJk5ltYrPV5FhPfSYVDJULqNCKrQCB4=\n"
+    "-----END RSA PRIVATE KEY-----\n";
+
+
+static const PKCS12_ATTR ATTRS1[] = {
+    { "friendlyName", "george" },
+    { "localKeyID", "1234567890" },
+    { "1.2.3.4.5", "MyCustomAttribute" },
+};
+
+static const PKCS12_ATTR ATTRS2[] = {
+    { "friendlyName", "janet" },
+    { "localKeyID", "987654321" },
+    { "1.2.3.5.8.13", "AnotherCustomAttribute" },
+};
+
+static const PKCS12_ENC enc_default = {
+    NID_pbe_WithSHA1And3_Key_TripleDES_CBC,
+    "Password1",
+    1000
+};
+
+static const PKCS12_ENC mac_default = {
+    NID_sha1,
+    "Password1",
+    1000
+};
+
+
+/* --------------------------------------------------------------------------
+ * PKCS12 builder functions
+ */
+
+X509 *load_cert(const char *cert_pem)
+{
+    X509 *cert;
+    BIO *mem = BIO_new(BIO_s_mem());
+    BIO_puts(mem, cert_pem);
+    cert = PEM_read_bio_X509(mem, NULL, 0, NULL);
+    BIO_free(mem);
+    return cert;
+}
+
+EVP_PKEY *load_pkey(const char *key_pem)
+{
+    EVP_PKEY *pkey;
+    BIO *mem = BIO_new(BIO_s_mem());
+    BIO_puts(mem, key_pem);
+    pkey = PEM_read_bio_PrivateKey(mem, NULL, 0, NULL);
+    BIO_free(mem);
+    return pkey;
+}
+
+
+int add_attributes(PKCS12_SAFEBAG *bag, const PKCS12_ATTR *attrs, int num_attrs)
+{
+    int i;
+    int attr_nid;
+
+    for (i = 0; i < num_attrs; i++) {
+        TEST_info("Adding attribute %s = %s", attrs[i].oid, attrs[i].value);
+        attr_nid = OBJ_txt2nid(attrs[i].oid);
+
+        if (attr_nid == NID_friendlyName) {
+            if (!TEST_true(PKCS12_add_friendlyname(bag, attrs[i].value, -1)))
+                goto err;
+        } else if (attr_nid == NID_localKeyID) {
+            if (!TEST_true(PKCS12_add_localkeyid(bag, (unsigned char *)attrs[i].value, strlen(attrs[i].value))))
+                goto err;
+        } else {
+            /* Custom attribute values limited to ASCII in these tests */
+            if (!TEST_true(PKCS12_add1_attr_by_txt(bag, attrs[i].oid, MBSTRING_ASC, (unsigned char *)attrs[i].value, strlen(attrs[i].value))))
+                goto err;
+        }
+    }
+    return 1;
+err:
+    return 0;
+}
+
+int generate_certbag(STACK_OF(PKCS12_SAFEBAG) **bags, const char *cert_pem, const PKCS12_ATTR *attrs, int num_attrs)
+{
+    PKCS12_SAFEBAG *bag = NULL;
+    X509 *cert = NULL;
+    char *name;
+
+    cert = load_cert(cert_pem);
+    if (!TEST_ptr(cert))
+        goto err;
+
+    name = X509_NAME_oneline(X509_get_subject_name(cert), NULL, 0);
+    TEST_info("Adding certificate <%s>", name);
+    OPENSSL_free(name);
+
+    bag = PKCS12_add_cert(bags, cert);
+    if (!TEST_ptr(bag))
+        goto err;
+
+    if (!TEST_true(add_attributes(bag, attrs, num_attrs)))
+        goto err;
+
+    X509_free(cert);
+    return 1;
+err:
+    X509_free(cert);
+    return 0;
```

Comment:
I would prefer to see the pattern used commonly elsewhere.. i.e:
int ret = 0;


   ret = 1;
err:
   X509_free(cert);
   return ret;

Reply:
You do this a bit in this file..

Reply:
Yeah this looks like a better pattern. Done.